File: javascript/345.reverse-vowels-of-a-string.js
/*
 * @lc app=leetcode id=345 lang=javascript
 *
 * [345] Reverse Vowels of a String
 */

// @lc code=start
/**
 * @param {string} s
 * @return {string}
 */
var reverseVowels = function (s) {
  // Solution 1:
  // const vowels = s.match(/[aeiouAEIOU]/g);
  // return s.replace(/[aeiouAEIOU]/g, (el) => vowels.pop());

  // Solution 2:
  const isVowel = (c) =>
    c === "a" ||
    c === "e" ||
    c === "i" ||
    c === "o" ||
    c === "u" ||
    c === "A" ||
    c === "E" ||
    c === "I" ||
    c === "O" ||
    c === "U";
  s = s.split("");
  let left = 0,
    right = s.length - 1;
  while (left < right) {
    while (left < right && !isVowel(s[left])) {
      left++;
    }
    while (left < right && !isVowel(s[right])) {
      right--;
    }
    if (left < right) {
      [s[left], s[right]] = [s[right], s[left]];
      left++;
      right--;
    }
  }
  return s.join("");
};
// @lc code=end

File: rust/345.reverse-vowels-of-a-string.rs
/*
 * @lc app=leetcode id=345 lang=rust
 *
 * [345] Reverse Vowels of a String
 */

// @lc code=start
impl Solution {
    pub fn reverse_vowels(s: String) -> String {
        let mut s: Vec<char> = s.chars().collect();
        let (mut left, mut right) = (0, s.len() - 1);
        let vowels = vec!['a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U'];
        while left < right {
            while left < right && !vowels.contains(&s[left]) {
                left += 1;
            }
            while left < right && !vowels.contains(&s[right]) {
                right -= 1;
            }
            if left < right {
                s.swap(left, right);
                left += 1;
                right -= 1;
            }
        }
        return s.iter().collect();
    }
}
// @lc code=end
